File: src/data/courses/dsa/trie/prefix-suffix/word-filter/code/java/WordFilter.java
// ============================================================
// Solution 1: Brute Force — Linear scan all words O(N*L)
// ============================================================
import java.util.*;
class Solution1 {
    List<String> words = new ArrayList<>();
    public void addWord(String word) { words.add(word); }
    public List<String> search(String query) {
        List<String> result = new ArrayList<>();
        for (String w : words) {
            boolean match = w.length() == query.length();
            if (match) for (int i = 0; i < query.length(); i++)
                if (query.charAt(i) != '.' && query.charAt(i) != w.charAt(i)) { match = false; break; }
            if (match) result.add(w);
        }
        return result;
    }
}
// ============================================================
// Solution 2: Optimal (Trie-based)
// ============================================================
: Optimal (Trie-based)
// ============================================================
import java.util.*;

// Time: O(N*L²) build, O(P+S) query | Space: O(N*L²)

class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    int weight = -1;
}

class WordFilter {
    private TrieNode root;
    
    public WordFilter(String[] words) {
        root = new TrieNode();
        for (int weight = 0; weight < words.length; weight++) {
            String word = words[weight];
            for (int i = 0; i <= word.length(); i++) {
                String suffix = word.substring(i);
                String key = suffix + "#" + word;
                insert(key, weight);
            }
        }
    }
    
    private void insert(String word, int weight) {
        TrieNode node = root;
        for (char ch : word.toCharArray()) {
            node.children.putIfAbsent(ch, new TrieNode());
            node = node.children.get(ch);
            node.weight = weight;
        }
    }
    
    public int f(String prefix, String suffix) {
        TrieNode node = root;
        String searchKey = suffix + "#" + prefix;
        for (char ch : searchKey.toCharArray()) {
            if (!node.children.containsKey(ch)) return -1;
            node = node.children.get(ch);
        }
        return node.weight;
    }
}

File: src/data/courses/dsa/trie/prefix-suffix/word-filter/code/cpp/WordFilter.cpp
// ============================================================
// Solution 1: Linear scan for prefix+suffix match
// Time: O(N * L) per query | Space: O(N * L)
// ============================================================
#include <vector>
#include <string>
using namespace std;

class Solution1 {
    vector<string> words;
public:
    Solution1(vector<string>& w) : words(w) {}
    int f(string prefix, string suffix) {
        for (int i = words.size() - 1; i >= 0; i--) {
            string& w = words[i];
            if (w.size() >= prefix.size() && w.size() >= suffix.size() &&
                w.substr(0, prefix.size()) == prefix &&
                w.substr(w.size() - suffix.size()) == suffix)
                return i;
        }
        return -1;  // O(N*L) per query
    }
};

// ============================================================
// Solution 2: Optimal (Trie-based)
// ============================================================
#include <vector>
#include <string>
#include <unordered_map>
using namespace std;

// Time: O(N*L²) build, O(P+S) query | Space: O(N*L²)

struct TrieNode {
    unordered_map<char, TrieNode*> children;
    int weight = -1;
};

class Solution2 {
    TrieNode* root;
    
    void insert(const string& word, int weight) {
        TrieNode* node = root;
        for (char ch : word) {
            if (!node->children[ch]) {
                node->children[ch] = new TrieNode();
            }
            node = node->children[ch];
            node->weight = weight;
        }
    }
    
public:
    WordFilter(vector<string>& words) {
        root = new TrieNode();
        for (int weight = 0; weight < words.size(); weight++) {
            string word = words[weight];
            for (int i = 0; i <= word.size(); i++) {
                string suffix = word.substr(i);
                string key = suffix + "#" + word;
                insert(key, weight);
            }
        }
    }
    
    int f(string prefix, string suffix) {
        TrieNode* node = root;
        string searchKey = suffix + "#" + prefix;
        for (char ch : searchKey) {
            if (!node->children[ch]) return -1;
            node = node->children[ch];
        }
        return node->weight;
    }
};
